server.py

```diff
@@ -141,13 +169,13 @@ def merge(response):
     try:
         db.session.merge(response)
         db.session.commit()
-    except SQLAlchemyError as e:
-        logger.info("Unable to save response", error=e)
-        raise SQLAlchemyError
     except IntegrityError as e:
-        logger.info("Integrity error in database. Rolling back commit",
-                    error=e)
-        raise IntegrityError
+        logger.error("Integrity error in database. Rolling back commit",
+                     error=e)
+        raise e
```

Comment:
You need to add explicit rollbacks in all the except blocks where you're wrapping a commit

 session.rollback()